--- zig-x86_64-linux-0.16.0-dev.3039+b490412cd/lib/std/Io/Threaded.zig
+++ zig-x86_64-linux-0.16.0-dev.3039+b490412cd/lib/std/Io/Threaded.zig
@@ -12069,19 +12069,174 @@
     options: IpAddress.ConnectOptions,
 ) IpAddress.ConnectError!net.Socket {
     if (!have_networking) return error.NetworkDown;
-    if (options.timeout != .none) @panic("TODO implement netConnectIpPosix with timeout");
     const t: *Threaded = @ptrCast(@alignCast(userdata));
-    _ = t;
     const family = posixAddressFamily(address);
     const socket_fd = try openSocketPosix(family, .{ .mode = options.mode, .protocol = options.protocol });
     errdefer closeFd(socket_fd);
     var storage: PosixAddress = undefined;
     var addr_len = addressToPosix(address, &storage);
-    try posixConnect(socket_fd, &storage.any, addr_len);
+    if (options.timeout == .none) {
+        try posixConnect(socket_fd, &storage.any, addr_len);
+    } else {
+        try posixConnectWithTimeout(t, socket_fd, &storage.any, addr_len, options.timeout);
+    }
     try posixGetSockName(socket_fd, &storage.any, &addr_len);
     return .{ .handle = socket_fd, .address = addressFromPosix(&storage) };
 }
 
+fn posixConnectWithTimeout(
+    t: *Threaded,
+    socket_fd: posix.socket_t,
+    addr: *const posix.sockaddr,
+    addr_len: posix.socklen_t,
+    timeout: Io.Timeout,
+) IpAddress.ConnectError!void {
+    const original_flags = try posixGetStatusFlags(socket_fd);
+    var nonblocking_flags = original_flags;
+    nonblocking_flags |= @as(usize, 1 << @bitOffsetOf(posix.O, "NONBLOCK"));
+    const need_restore_flags = nonblocking_flags != original_flags;
+    if (need_restore_flags) try posixSetStatusFlags(socket_fd, nonblocking_flags);
+
+    const result: IpAddress.ConnectError!void = r: {
+        posixConnect(socket_fd, addr, addr_len) catch |err| switch (err) {
+            error.WouldBlock, error.ConnectionPending => {},
+            else => |e| break :r e,
+        };
+
+        const t_io = io(t);
+        const deadline = timeout.toTimestamp(t_io);
+        var poll_fd: [1]posix.pollfd = .{.{
+            .fd = socket_fd,
+            .events = posix.POLL.OUT | posix.POLL.ERR,
+            .revents = 0,
+        }};
+
+        while (true) {
+            const timeout_ms: i32 = t: {
+                const d = deadline orelse break :t -1;
+                const duration = d.durationFromNow(t_io);
+                break :t @min(@max(0, duration.raw.toMilliseconds()), std.math.maxInt(i32));
+            };
+            const syscall = try Syscall.start();
+            const rc = posix.system.poll(&poll_fd, 1, timeout_ms);
+            switch (posix.errno(rc)) {
+                .SUCCESS => {
+                    syscall.finish();
+                    if (rc == 0) {
+                        if (deadline == null) continue;
+                        break :r error.Timeout;
+                    }
+                    posixConnectCheckSoError(socket_fd) catch |err| switch (err) {
+                        error.WouldBlock, error.ConnectionPending => continue,
+                        else => |e| break :r e,
+                    };
+                    break :r;
+                },
+                .INTR => {
+                    try syscall.checkCancel();
+                    continue;
+                },
+                .BADF => break :r syscall.errnoBug(.BADF),
+                .FAULT => break :r syscall.errnoBug(.FAULT),
+                .INVAL => break :r syscall.errnoBug(.INVAL),
+                else => |e| break :r syscall.unexpectedErrno(e),
+            }
+        }
+    };
+    if (need_restore_flags) try posixSetStatusFlags(socket_fd, original_flags);
+    return result;
+}
+
+fn posixConnectCheckSoError(socket_fd: posix.socket_t) IpAddress.ConnectError!void {
+    var so_error: c_int = 0;
+    var opt_len: posix.socklen_t = @sizeOf(c_int);
+    const syscall: Syscall = try .start();
+    while (true) {
+        switch (posix.errno(posix.system.getsockopt(
+            socket_fd,
+            posix.SOL.SOCKET,
+            posix.SO.ERROR,
+            @ptrCast(&so_error),
+            &opt_len,
+        ))) {
+            .SUCCESS => {
+                syscall.finish();
+                if (so_error == 0) return;
+                const connect_errno: posix.E = @enumFromInt(so_error);
+                return switch (connect_errno) {
+                    .ADDRNOTAVAIL => error.AddressUnavailable,
+                    .AFNOSUPPORT => error.AddressFamilyUnsupported,
+                    .AGAIN, .INPROGRESS => error.WouldBlock,
+                    .ALREADY => error.ConnectionPending,
+                    .CONNREFUSED => error.ConnectionRefused,
+                    .CONNRESET => error.ConnectionResetByPeer,
+                    .HOSTUNREACH => error.HostUnreachable,
+                    .NETUNREACH => error.NetworkUnreachable,
+                    .TIMEDOUT => error.Timeout,
+                    .ACCES => error.AccessDenied,
+                    .NETDOWN => error.NetworkDown,
+                    .BADF => |err| errnoBug(err),
+                    .CONNABORTED => |err| errnoBug(err),
+                    .FAULT => |err| errnoBug(err),
+                    .ISCONN => |err| errnoBug(err),
+                    .NOENT => |err| errnoBug(err),
+                    .NOTSOCK => |err| errnoBug(err),
+                    .PERM => |err| errnoBug(err),
+                    .PROTOTYPE => |err| errnoBug(err),
+                    else => |err| posix.unexpectedErrno(err),
+                };
+            },
+            .INTR => {
+                try syscall.checkCancel();
+                continue;
+            },
+            .BADF => |err| return syscall.errnoBug(err),
+            .FAULT => |err| return syscall.errnoBug(err),
+            .INVAL => |err| return syscall.errnoBug(err),
+            .NOTSOCK => |err| return syscall.errnoBug(err),
+            else => |err| return syscall.unexpectedErrno(err),
+        }
+    }
+}
+
+fn posixGetStatusFlags(fd: posix.fd_t) IpAddress.ConnectError!usize {
+    const syscall: Syscall = try .start();
+    while (true) {
+        const rc = posix.system.fcntl(fd, posix.F.GETFL, @as(usize, 0));
+        switch (posix.errno(rc)) {
+            .SUCCESS => {
+                syscall.finish();
+                return @intCast(rc);
+            },
+            .INTR => {
+                try syscall.checkCancel();
+                continue;
+            },
+            .BADF => |err| return syscall.errnoBug(err),
+            else => |err| return syscall.unexpectedErrno(err),
+        }
+    }
+}
+
+fn posixSetStatusFlags(fd: posix.fd_t, flags: usize) IpAddress.ConnectError!void {
+    const syscall: Syscall = try .start();
+    while (true) {
+        switch (posix.errno(posix.system.fcntl(fd, posix.F.SETFL, flags))) {
+            .SUCCESS => {
+                syscall.finish();
+                return;
+            },
+            .INTR => {
+                try syscall.checkCancel();
+                continue;
+            },
+            .BADF => |err| return syscall.errnoBug(err),
+            .INVAL => |err| return syscall.errnoBug(err),
+            else => |err| return syscall.unexpectedErrno(err),
+        }
+    }
+}
+
 fn netConnectIpWindows(
     userdata: ?*anyopaque,
     address: *const IpAddress,
@@ -13034,34 +13189,40 @@
         .flags = 0,
     };
     addBuf(&iovecs, &msg.iovlen, header);
-    for (data[0 .. data.len - 1]) |bytes| addBuf(&iovecs, &msg.iovlen, bytes);
-    const pattern = data[data.len - 1];
+    if (data.len > 0) {
+        for (data[0 .. data.len - 1]) |bytes| addBuf(&iovecs, &msg.iovlen, bytes);
+    } else {
+        assert(splat == 0);
+    }
 
     var splat_backup_buffer: [splat_buffer_size]u8 = undefined;
-    if (iovecs.len - msg.iovlen != 0) switch (splat) {
-        0 => {},
-        1 => addBuf(&iovecs, &msg.iovlen, pattern),
-        else => switch (pattern.len) {
+    if (data.len > 0 and iovecs.len - msg.iovlen != 0) {
+        const pattern = data[data.len - 1];
+        switch (splat) {
             0 => {},
-            1 => {
-                const splat_buffer = &splat_backup_buffer;
-                const memset_len = @min(splat_buffer.len, splat);
-                const buf = splat_buffer[0..memset_len];
-                @memset(buf, pattern[0]);
-                addBuf(&iovecs, &msg.iovlen, buf);
-                var remaining_splat = splat - buf.len;
-                while (remaining_splat > splat_buffer.len and iovecs.len - msg.iovlen != 0) {
-                    assert(buf.len == splat_buffer.len);
-                    addBuf(&iovecs, &msg.iovlen, splat_buffer);
-                    remaining_splat -= splat_buffer.len;
-                }
-                addBuf(&iovecs, &msg.iovlen, splat_buffer[0..@min(remaining_splat, splat_buffer.len)]);
-            },
-            else => for (0..@min(splat, iovecs.len - msg.iovlen)) |_| {
-                addBuf(&iovecs, &msg.iovlen, pattern);
+            1 => addBuf(&iovecs, &msg.iovlen, pattern),
+            else => switch (pattern.len) {
+                0 => {},
+                1 => {
+                    const splat_buffer = &splat_backup_buffer;
+                    const memset_len = @min(splat_buffer.len, splat);
+                    const buf = splat_buffer[0..memset_len];
+                    @memset(buf, pattern[0]);
+                    addBuf(&iovecs, &msg.iovlen, buf);
+                    var remaining_splat = splat - buf.len;
+                    while (remaining_splat > splat_buffer.len and iovecs.len - msg.iovlen != 0) {
+                        assert(buf.len == splat_buffer.len);
+                        addBuf(&iovecs, &msg.iovlen, splat_buffer);
+                        remaining_splat -= splat_buffer.len;
+                    }
+                    addBuf(&iovecs, &msg.iovlen, splat_buffer[0..@min(remaining_splat, splat_buffer.len)]);
+                },
+                else => for (0..@min(splat, iovecs.len - msg.iovlen)) |_| {
+                    addBuf(&iovecs, &msg.iovlen, pattern);
+                },
             },
-        },
-    };
+        }
+    }
     const flags = posix.MSG.NOSIGNAL;
 
     const syscall: Syscall = try .start();
 --- zig-x86_64-linux-0.16.0-dev.3039+b490412cd/lib/std/Io/Uring.zig
 +++ zig-x86_64-linux-0.16.0-dev.3039+b490412cd/lib/std/Io/Uring.zig
 @@ -771,16 +771,16 @@
             .random = random,
             .randomSecure = randomSecure,
 
-            .netListenIp = netListenIpUnavailable,
-            .netAccept = netAcceptUnavailable,
+            .netListenIp = netListenIp,
+            .netAccept = netAccept,
             .netBindIp = netBindIp,
-            .netConnectIp = netConnectIpUnavailable,
+            .netConnectIp = netConnectIp,
             .netListenUnix = netListenUnixUnavailable,
             .netConnectUnix = netConnectUnixUnavailable,
             .netSocketCreatePair = netSocketCreatePairUnavailable,
-            .netSend = netSendUnavailable,
-            .netRead = netReadUnavailable,
-            .netWrite = netWriteUnavailable,
+            .netSend = netSend,
+            .netRead = netRead,
+            .netWrite = netWrite,
             .netWriteFile = netWriteFileUnavailable,
             .netClose = netClose,
             .netShutdown = netShutdown,
@@ -1215,7 +1215,18 @@
                    },
                    0b10 => {
                        const batch_userdata: *Io.Operation.Storage.Pending.Userdata =
                            @ptrFromInt(cqe.user_data & ~@as(usize, 0b11));
+                        if (batch_userdata[0] == 0) break :ready_fiber null;
                        const batch: *Io.Batch = @ptrFromInt(batch_userdata[0]);
                        var next: usize = 0b00;
                        batch_userdata[0..3].* = .{ next, @as(u32, @bitCast(cqe.res)), cqe.flags };
                        while (true) {
                            next = @cmpxchgWeak(
                                usize,
                                @as(*usize, @ptrCast(&batch.userdata)),
                                next,
                                cqe.user_data,
                                .release,
                                .acquire,
                            ) orelse break;
                            batch_userdata[0] = next;
                        }
                        break :ready_fiber switch (@as(u2, @truncate(next))) {
                            0b00, 0b01 => @ptrFromInt(next & ~@as(usize, 0b11)),
                            0b10, 0b11 => null,
                        };
                    },
 @@ -2308,6 +2308,192 @@
     }
 }
 
+const BatchNetReceiveContext = struct {
+    active_count: usize,
+    states: []BatchNetReceiveState,
+};
+
+const BatchNetReceiveState = struct {
+    active: bool,
+    socket_handle: net.Socket.Handle,
+    flags: net.ReceiveFlags,
+    message_buffer: []net.IncomingMessage,
+    data_buffer: []u8,
+    message_count: usize,
+    data_used: usize,
+    storage: PosixAddress,
+    iov: iovec,
+    msg: linux.msghdr,
+};
+
+fn batchNetReceiveContextCreate(ev: *Evented, len: usize) Allocator.Error!*BatchNetReceiveContext {
+    const allocator = ev.allocator();
+    const context = try allocator.create(BatchNetReceiveContext);
+    errdefer allocator.destroy(context);
+    const states = try allocator.alloc(BatchNetReceiveState, len);
+    errdefer allocator.free(states);
+    for (states) |*state| state.* = .{
+        .active = false,
+        .socket_handle = undefined,
+        .flags = .{},
+        .message_buffer = &.{},
+        .data_buffer = &.{},
+        .message_count = 0,
+        .data_used = 0,
+        .storage = undefined,
+        .iov = undefined,
+        .msg = undefined,
+    };
+    context.* = .{
+        .active_count = 0,
+        .states = states,
+    };
+    return context;
+}
+
+fn batchNetReceiveContextDestroy(ev: *Evented, context: *BatchNetReceiveContext) void {
+    const allocator = ev.allocator();
+    allocator.free(context.states);
+    allocator.destroy(context);
+}
+
+fn batchNetReceiveContextRelease(
+    ev: *Evented,
+    context: *BatchNetReceiveContext,
+    state: *BatchNetReceiveState,
+) void {
+    assert(state.active);
+    assert(context.active_count > 0);
+    state.active = false;
+    state.message_buffer = &.{};
+    state.data_buffer = &.{};
+    context.active_count -= 1;
+    if (context.active_count == 0) batchNetReceiveContextDestroy(ev, context);
+}
+
+fn netReceiveRwFlags(flags: net.ReceiveFlags) u32 {
+    return linux.MSG.NOSIGNAL |
+        @as(u32, if (flags.oob) linux.MSG.OOB else 0) |
+        @as(u32, if (flags.peek) linux.MSG.PEEK else 0) |
+        @as(u32, if (flags.trunc) linux.MSG.TRUNC else 0);
+}
+
+fn initBatchNetReceiveState(state: *BatchNetReceiveState, o: Io.Operation.NetReceive) void {
+    assert(o.message_buffer.len > 0);
+    assert(o.data_buffer.len > 0);
+    state.* = .{
+        .active = true,
+        .socket_handle = o.socket_handle,
+        .flags = o.flags,
+        .message_buffer = o.message_buffer,
+        .data_buffer = o.data_buffer,
+        .message_count = 0,
+        .data_used = 0,
+        .storage = undefined,
+        .iov = undefined,
+        .msg = undefined,
+    };
+}
+
+fn prepareBatchNetReceiveMsg(state: *BatchNetReceiveState) void {
+    assert(state.active);
+    assert(state.message_count < state.message_buffer.len);
+    assert(state.data_used < state.data_buffer.len);
+    const message = &state.message_buffer[state.message_count];
+    const remaining_data_buffer = state.data_buffer[state.data_used..];
+    state.iov = .{
+        .base = remaining_data_buffer.ptr,
+        .len = remaining_data_buffer.len,
+    };
+    state.msg = .{
+        .name = &state.storage.any,
+        .namelen = @sizeOf(PosixAddress),
+        .iov = (&state.iov)[0..1],
+        .iovlen = 1,
+        .control = if (message.control.len == 0) null else message.control.ptr,
+        .controllen = @intCast(message.control.len),
+        .flags = undefined,
+    };
+}
+
+fn finishBatchNetReceiveMessage(state: *BatchNetReceiveState, result_len: usize) void {
+    assert(state.active);
+    assert(state.message_count < state.message_buffer.len);
+    assert(state.data_used <= state.data_buffer.len);
+    const message = &state.message_buffer[state.message_count];
+    const remaining_data_buffer = state.data_buffer[state.data_used..];
+    const used_len = @min(result_len, remaining_data_buffer.len);
+    const data = remaining_data_buffer[0..used_len];
+    state.data_used += data.len;
+    message.* = .{
+        .from = addressFromPosix(&state.storage),
+        .data = data,
+        .control = if (state.msg.control) |ptr|
+            @as([*]u8, @ptrCast(ptr))[0..state.msg.controllen]
+        else
+            message.control,
+        .flags = .{
+            .eor = state.msg.flags & linux.MSG.EOR != 0,
+            .trunc = state.msg.flags & linux.MSG.TRUNC != 0,
+            .ctrunc = state.msg.flags & linux.MSG.CTRUNC != 0,
+            .oob = state.msg.flags & linux.MSG.OOB != 0,
+            .errqueue = state.msg.flags & linux.MSG.ERRQUEUE != 0,
+        },
+    };
+    state.message_count += 1;
+}
+
+fn batchNetReceiveTerminalResult(
+    err: linux.E,
+    message_count: usize,
+) struct { ?net.Socket.ReceiveError, usize } {
+    return .{ switch (err) {
+        .INTR => errnoBug(.INTR),
+        .BADF => errnoBug(.BADF),
+        .NFILE => error.SystemFdQuotaExceeded,
+        .MFILE => error.ProcessFdQuotaExceeded,
+        .FAULT => errnoBug(.FAULT),
+        .INVAL => errnoBug(.INVAL),
+        .NOBUFS => error.SystemResources,
+        .NOMEM => error.SystemResources,
+        .NOTCONN => error.SocketUnconnected,
+        .NOTSOCK => errnoBug(.NOTSOCK),
+        .MSGSIZE => error.MessageOversize,
+        .PIPE => error.SocketUnconnected,
+        .OPNOTSUPP => errnoBug(.OPNOTSUPP),
+        .CONNRESET => error.ConnectionResetByPeer,
+        .NETDOWN => error.NetworkDown,
+        else => unexpectedErrno(err),
+    }, message_count };
+}
+
+fn drainBatchNetReceiveReady(
+    state: *BatchNetReceiveState,
+) struct { ?net.Socket.ReceiveError, usize } {
+    assert(state.active);
+    while (state.message_count < state.message_buffer.len and
+        state.data_used < state.data_buffer.len and
+        !(state.flags.peek and state.message_count != 0))
+    {
+        prepareBatchNetReceiveMsg(state);
+        const rc = linux.recvmsg(
+            state.socket_handle,
+            &state.msg,
+            netReceiveRwFlags(state.flags) | linux.MSG.DONTWAIT,
+        );
+        switch (linux.errno(rc)) {
+            .SUCCESS => {
+                finishBatchNetReceiveMessage(state, @intCast(rc));
+                continue;
+            },
+            .INTR => continue,
+            .AGAIN => return .{ null, state.message_count },
+            else => |err| return batchNetReceiveTerminalResult(err, state.message_count),
+        }
+    }
+    return .{ null, state.message_count };
+}
+
 /// If `concurrency` is false, `error.ConcurrencyUnavailable` is unreachable.
 fn batchDrainSubmitted(
     ev: *Evented,
@@ -2318,6 +2504,10 @@
     var index = batch.submitted.head;
     if (index == .none) return;
     const thread = try maybe_sync.cancelRegion().awaitIoUring();
+    var net_receive_context: ?*BatchNetReceiveContext = null;
+    errdefer if (net_receive_context) |context| {
+        if (context.active_count == 0) batchNetReceiveContextDestroy(ev, context);
+    };
     errdefer batch.submitted.head = index;
     while (index != .none) {
         const storage = &batch.storage[index.toIndex()];
@@ -2388,9 +2578,62 @@
                 return error.ConcurrencyUnavailable
             else
                 .{ .device_io_control = try ev.deviceIoControl(try maybe_sync.enterSync(ev), o) },
-            .net_receive => |o| {
-                _ = o;
-                @panic("TODO implement batchDrainSubmitted for net_receive");
+            .net_receive => |o| if (concurrency) {
+                if (o.message_buffer.len == 0 or o.data_buffer.len == 0) {
+                    break :result .{ .net_receive = .{ null, 0 } };
+                }
+                const context = net_receive_context orelse context: {
+                    const context = batchNetReceiveContextCreate(ev, batch.storage.len) catch
+                        return error.ConcurrencyUnavailable;
+                    net_receive_context = context;
+                    break :context context;
+                };
+                const state = &context.states[index.toIndex()];
+                assert(!state.active);
+                initBatchNetReceiveState(state, o);
+                context.active_count += 1;
+                errdefer if (state.active) batchNetReceiveContextRelease(ev, context, state);
+                prepareBatchNetReceiveMsg(state);
+                storage.* = .{ .pending = .{
+                    .node = .{ .prev = batch.pending.tail, .next = .none },
+                    .tag = .net_receive,
+                    .userdata = undefined,
+                } };
+                storage.pending.userdata[3] = @intFromPtr(context);
+                thread.enqueue().* = .{
+                    .opcode = .RECVMSG,
+                    .flags = 0,
+                    .ioprio = 0,
+                    .fd = state.socket_handle,
+                    .off = 0,
+                    .addr = @intFromPtr(&state.msg),
+                    .len = 0,
+                    .rw_flags = netReceiveRwFlags(state.flags),
+                    .user_data = @intFromPtr(&storage.pending.userdata) | 0b10,
+                    .buf_index = 0,
+                    .personality = 0,
+                    .splice_fd_in = 0,
+                    .addr3 = 0,
+                    .resv = 0,
+                };
+                break :result null;
+            } else .{
+                .net_receive = r: {
+                    const opt_err, const n = ev.netReceive(
+                        &maybe_sync.cancel_region,
+                        o.socket_handle,
+                        o.message_buffer,
+                        o.data_buffer,
+                        o.flags,
+                    );
+                    break :r .{
+                        if (opt_err) |err| switch (err) {
+                            error.Canceled => |e| return e,
+                            else => |e| e,
+                        } else null,
+                        n,
+                    };
+                },
             },
         })) |result| {
             switch (batch.completed.tail) {
@@ -2492,7 +2735,25 @@
                     },
                 },
                 .device_io_control => unreachable,
-                .net_receive => @panic("TODO"),
+                .net_receive => net_receive_result: {
+                    const context: *BatchNetReceiveContext = @ptrFromInt(pending.userdata[3]);
+                    const state = &context.states[index.toIndex()];
+                    assert(state.active);
+                    defer batchNetReceiveContextRelease(ev, context, state);
+                    break :net_receive_result switch (completion.errno()) {
+                        .SUCCESS => .{
+                            .net_receive = receive_result: {
+                                finishBatchNetReceiveMessage(state, @intCast(completion.result));
+                                break :receive_result drainBatchNetReceiveReady(state);
+                            },
+                        },
+                        .CANCELED => break :net_receive_result null,
+                        else => .{ .net_receive = batchNetReceiveTerminalResult(
+                            completion.errno(),
+                            state.message_count,
+                        ) },
+                    };
+                },
             })) |result| {
                 switch (batch.completed.tail) {
                     .none => batch.completed.head = index,
@@ -4953,28 +5214,88 @@
     };
 }
 
-fn netListenIpUnavailable(
+fn netListenIp(
     userdata: ?*anyopaque,
     address: *const net.IpAddress,
     options: net.IpAddress.ListenOptions,
 ) net.IpAddress.ListenError!net.Socket {
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = address;
-    _ = options;
-    return error.NetworkDown;
+    const family = posixAddressFamily(address);
+    var maybe_sync: CancelRegion.Sync.Maybe = .{ .cancel_region = .init() };
+    defer maybe_sync.deinit(ev);
+    const socket_fd = try ev.socket(&maybe_sync.cancel_region, family, .{
+        .mode = options.mode,
+        .protocol = options.protocol,
+    });
+    errdefer ev.closeAsync(socket_fd);
+
+    if (options.reuse_address) {
+        try ev.setsockopt(&maybe_sync.cancel_region, socket_fd, linux.SOL.SOCKET, linux.SO.REUSEADDR, 1);
+        if (@hasDecl(linux.SO, "REUSEPORT")) {
+            try ev.setsockopt(&maybe_sync.cancel_region, socket_fd, linux.SOL.SOCKET, linux.SO.REUSEPORT, 1);
+        }
+    }
+
+    var storage: PosixAddress = undefined;
+    var addr_len = addressToPosix(address, &storage);
+    try ev.bind(&maybe_sync.cancel_region, socket_fd, &storage.any, addr_len);
+    try ev.listen(&maybe_sync.cancel_region, socket_fd, options.kernel_backlog);
+    try ev.getsockname(try maybe_sync.enterSync(ev), socket_fd, &storage.any, &addr_len);
+    return .{ .handle = socket_fd, .address = addressFromPosix(&storage) };
 }
 
-fn netAcceptUnavailable(
+fn netAccept(
     userdata: ?*anyopaque,
     listen_handle: net.Socket.Handle,
     options: net.Server.AcceptOptions,
 ) net.Server.AcceptError!net.Socket {
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = listen_handle;
     _ = options;
-    return error.NetworkDown;
+    var cancel_region: CancelRegion = .init();
+    defer cancel_region.deinit();
+    var storage: PosixAddress = undefined;
+    var addr_len: linux.socklen_t = @sizeOf(PosixAddress);
+
+    while (true) {
+        const thread = try cancel_region.awaitIoUring();
+        thread.enqueue().* = .{
+            .opcode = .ACCEPT,
+            .flags = 0,
+            .ioprio = 0,
+            .fd = listen_handle,
+            .off = @intFromPtr(&addr_len),
+            .addr = @intFromPtr(&storage.any),
+            .len = 0,
+            .rw_flags = linux.SOCK.CLOEXEC,
+            .user_data = @intFromPtr(cancel_region.fiber),
+            .buf_index = 0,
+            .personality = 0,
+            .splice_fd_in = 0,
+            .addr3 = 0,
+            .resv = 0,
+        };
+        ev.yield(null, .nothing);
+        const completion = cancel_region.completion();
+        switch (completion.errno()) {
+            .SUCCESS => return .{ .handle = completion.result, .address = addressFromPosix(&storage) },
+            .INTR, .CANCELED => {},
+            .AGAIN => return error.WouldBlock,
+            .BADF => |err| return errnoBug(err),
+            .CONNABORTED => return error.ConnectionAborted,
+            .FAULT => |err| return errnoBug(err),
+            .INVAL => return error.SocketNotListening,
+            .NOTSOCK => |err| return errnoBug(err),
+            .MFILE => return error.ProcessFdQuotaExceeded,
+            .NFILE => return error.SystemFdQuotaExceeded,
+            .NOBUFS => return error.SystemResources,
+            .NOMEM => return error.SystemResources,
+            .OPNOTSUPP => |err| return errnoBug(err),
+            .PROTO => return error.ProtocolFailure,
+            .PERM => return error.BlockedByFirewall,
+            .NETDOWN => return error.NetworkDown,
+            else => |err| return unexpectedErrno(err),
+        }
+    }
 }
 
 fn netBindIp(
@@ -4996,16 +5317,27 @@
     return .{ .handle = socket_fd, .address = addressFromPosix(&storage) };
 }
 
-fn netConnectIpUnavailable(
+fn netConnectIp(
     userdata: ?*anyopaque,
     address: *const net.IpAddress,
     options: net.IpAddress.ConnectOptions,
 ) net.IpAddress.ConnectError!net.Socket {
+    if (options.timeout != .none) return error.OptionUnsupported;
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = address;
-    _ = options;
-    return error.NetworkDown;
+    const family = posixAddressFamily(address);
+    var maybe_sync: CancelRegion.Sync.Maybe = .{ .cancel_region = .init() };
+    defer maybe_sync.deinit(ev);
+    const socket_fd = try ev.socket(&maybe_sync.cancel_region, family, .{
+        .mode = options.mode,
+        .protocol = options.protocol,
+    });
+    errdefer ev.closeAsync(socket_fd);
+
+    var storage: PosixAddress = undefined;
+    var addr_len = addressToPosix(address, &storage);
+    try ev.connect(&maybe_sync.cancel_region, socket_fd, &storage.any, addr_len);
+    try ev.getsockname(try maybe_sync.enterSync(ev), socket_fd, &storage.any, &addr_len);
+    return .{ .handle = socket_fd, .address = addressFromPosix(&storage) };
 }
 
 fn netListenUnixUnavailable(
@@ -5039,18 +5371,98 @@
     return error.OperationUnsupported;
 }
 
-fn netSendUnavailable(
+fn netSend(
     userdata: ?*anyopaque,
     handle: net.Socket.Handle,
     messages: []net.OutgoingMessage,
     flags: net.SendFlags,
 ) struct { ?net.Socket.SendError, usize } {
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = handle;
-    _ = messages;
-    _ = flags;
-    return .{ error.NetworkDown, 0 };
+    const msg_flags: u32 =
+        @as(u32, if (@hasDecl(linux.MSG, "CONFIRM") and flags.confirm) linux.MSG.CONFIRM else 0) |
+        @as(u32, if (@hasDecl(linux.MSG, "DONTROUTE") and flags.dont_route) linux.MSG.DONTROUTE else 0) |
+        @as(u32, if (@hasDecl(linux.MSG, "EOR") and flags.eor) linux.MSG.EOR else 0) |
+        @as(u32, if (@hasDecl(linux.MSG, "OOB") and flags.oob) linux.MSG.OOB else 0) |
+        @as(u32, if (@hasDecl(linux.MSG, "FASTOPEN") and flags.fastopen) linux.MSG.FASTOPEN else 0) |
+        linux.MSG.NOSIGNAL;
+
+    var cancel_region: CancelRegion = .init();
+    defer cancel_region.deinit();
+    var i: usize = 0;
+    while (i < messages.len) : (i += 1) {
+        netSendOne(ev, &cancel_region, handle, &messages[i], msg_flags) catch |err| return .{ err, i };
+    }
+    return .{ null, i };
+}
+
+fn netSendOne(
+    ev: *Evented,
+    cancel_region: *CancelRegion,
+    handle: net.Socket.Handle,
+    message: *net.OutgoingMessage,
+    flags: u32,
+) net.Socket.SendError!void {
+    var addr: PosixAddress = undefined;
+    var iov: iovec_const = .{ .base = message.data_ptr, .len = message.data_len };
+    const msg: linux.msghdr_const = .{
+        .name = &addr.any,
+        .namelen = addressToPosix(message.address, &addr),
+        .iov = (&iov)[0..1],
+        .iovlen = 1,
+        .control = if (message.control.len == 0) null else message.control.ptr,
+        .controllen = message.control.len,
+        .flags = 0,
+    };
+
+    while (true) {
+        const thread = try cancel_region.awaitIoUring();
+        thread.enqueue().* = .{
+            .opcode = .SENDMSG,
+            .flags = 0,
+            .ioprio = 0,
+            .fd = handle,
+            .off = 0,
+            .addr = @intFromPtr(&msg),
+            .len = 1,
+            .rw_flags = flags,
+            .user_data = @intFromPtr(cancel_region.fiber),
+            .buf_index = 0,
+            .personality = 0,
+            .splice_fd_in = 0,
+            .addr3 = 0,
+            .resv = 0,
+        };
+        ev.yield(null, .nothing);
+        const completion = cancel_region.completion();
+        switch (completion.errno()) {
+            .SUCCESS => {
+                message.data_len = @intCast(completion.result);
+                return;
+            },
+            .INTR, .CANCELED => {},
+            .ACCES => return error.AccessDenied,
+            .ALREADY => return error.FastOpenAlreadyInProgress,
+            .BADF => |err| return errnoBug(err),
+            .CONNRESET => return error.ConnectionResetByPeer,
+            .DESTADDRREQ => |err| return errnoBug(err),
+            .FAULT => |err| return errnoBug(err),
+            .INVAL => |err| return errnoBug(err),
+            .ISCONN => |err| return errnoBug(err),
+            .MSGSIZE => return error.MessageOversize,
+            .NOBUFS => return error.SystemResources,
+            .NOMEM => return error.SystemResources,
+            .NOTSOCK => |err| return errnoBug(err),
+            .OPNOTSUPP => |err| return errnoBug(err),
+            .PIPE => return error.SocketUnconnected,
+            .AFNOSUPPORT => return error.AddressFamilyUnsupported,
+            .HOSTUNREACH => return error.HostUnreachable,
+            .NETUNREACH => return error.NetworkUnreachable,
+            .NOTCONN => return error.SocketUnconnected,
+            .NETDOWN => return error.NetworkDown,
+            .CONNREFUSED => return error.ConnectionRefused, // Valid for datagram sendmsg after ICMP unreachable.
+            else => |err| return unexpectedErrno(err),
+        }
+    }
 }
 
 fn netReceive(
@@ -5066,6 +5478,7 @@
 
     while (true) {
         if (message_buffer.len - message_i == 0) return .{ null, message_i };
+        if (data_buffer.len - data_i == 0) return .{ null, message_i };
         const message = &message_buffer[message_i];
         const remaining_data_buffer = data_buffer[data_i..];
         var storage: PosixAddress = undefined;
@@ -5089,10 +5502,7 @@
             .off = 0,
             .addr = @intFromPtr(&msg),
             .len = 0,
-            .rw_flags = linux.MSG.NOSIGNAL |
-                @as(u32, if (flags.oob) linux.MSG.OOB else 0) |
-                @as(u32, if (flags.peek) linux.MSG.PEEK else 0) |
-                @as(u32, if (flags.trunc) linux.MSG.TRUNC else 0),
+            .rw_flags = netReceiveRwFlags(flags),
             .user_data = @intFromPtr(cancel_region.fiber),
             .buf_index = 0,
             .personality = 0,
@@ -5104,7 +5514,8 @@
         const completion = cancel_region.completion();
         switch (completion.errno()) {
             .SUCCESS => {
-                const data = remaining_data_buffer[0..@intCast(completion.result)];
+                const data_len = @min(@as(usize, @intCast(completion.result)), remaining_data_buffer.len);
+                const data = remaining_data_buffer[0..data_len];
                 data_i += data.len;
                 message.* = .{
                     .from = addressFromPosix(&storage),
@@ -5119,6 +5530,7 @@
                     },
                 };
                 message_i += 1;
+                if (flags.peek) return .{ null, message_i };
                 continue;
             },
             .AGAIN => unreachable,
@@ -5142,19 +5554,97 @@
     }
 }
 
-fn netReadUnavailable(
+fn netWriteMsg(
+    ev: *Evented,
+    cancel_region: *CancelRegion,
+    handle: net.Socket.Handle,
+    msg: *const linux.msghdr_const,
+    flags: u32,
+) net.Stream.Writer.Error!usize {
+    while (true) {
+        const thread = try cancel_region.awaitIoUring();
+        thread.enqueue().* = .{
+            .opcode = .SENDMSG,
+            .flags = 0,
+            .ioprio = 0,
+            .fd = handle,
+            .off = 0,
+            .addr = @intFromPtr(msg),
+            .len = 1,
+            .rw_flags = flags,
+            .user_data = @intFromPtr(cancel_region.fiber),
+            .buf_index = 0,
+            .personality = 0,
+            .splice_fd_in = 0,
+            .addr3 = 0,
+            .resv = 0,
+        };
+        ev.yield(null, .nothing);
+        const completion = cancel_region.completion();
+        switch (completion.errno()) {
+            .SUCCESS => return @intCast(completion.result),
+            .INTR, .CANCELED => {},
+            .ACCES => |err| return errnoBug(err),
+            .AGAIN => |err| return errnoBug(err),
+            .ALREADY => return error.FastOpenAlreadyInProgress,
+            .BADF => |err| return errnoBug(err),
+            .CONNRESET => return error.ConnectionResetByPeer,
+            .DESTADDRREQ => |err| return errnoBug(err), // Stream socket is already connected.
+            .FAULT => |err| return errnoBug(err),
+            .INVAL => |err| return errnoBug(err),
+            .ISCONN => |err| return errnoBug(err),
+            .MSGSIZE => |err| return errnoBug(err),
+            .NOBUFS => return error.SystemResources,
+            .NOMEM => return error.SystemResources,
+            .NOTSOCK => |err| return errnoBug(err),
+            .OPNOTSUPP => |err| return errnoBug(err),
+            .PIPE => return error.SocketUnconnected,
+            .AFNOSUPPORT => return error.AddressFamilyUnsupported,
+            .HOSTUNREACH => return error.HostUnreachable,
+            .NETUNREACH => return error.NetworkUnreachable,
+            .NOTCONN => return error.SocketUnconnected,
+            .NETDOWN => return error.NetworkDown,
+            .CONNREFUSED => return error.ConnectionRefused,
+            else => |err| return unexpectedErrno(err),
+        }
+    }
+}
+
+fn netRead(
     userdata: ?*anyopaque,
     fd: net.Socket.Handle,
     data: [][]u8,
 ) net.Stream.Reader.Error!usize {
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = fd;
-    _ = data;
-    return error.NetworkDown;
+
+    var iovecs_buffer: [max_iovecs_len]iovec = undefined;
+    var i: usize = 0;
+    for (data) |buf| {
+        if (i >= iovecs_buffer.len) break;
+        if (buf.len == 0) continue;
+        iovecs_buffer[i] = .{ .base = buf.ptr, .len = buf.len };
+        i += 1;
+    }
+    if (i == 0) return 0;
+
+    var cancel_region: CancelRegion = .init();
+    defer cancel_region.deinit();
+    return ev.preadv(&cancel_region, fd, iovecs_buffer[0..i], null) catch |err| switch (err) {
+        error.Canceled => return error.Canceled,
+        error.AccessDenied => return error.AccessDenied,
+        error.ConnectionResetByPeer => return error.ConnectionResetByPeer,
+        error.SocketUnconnected => return error.SocketUnconnected,
+        error.SystemResources => return error.SystemResources,
+        error.WouldBlock => return errnoBug(.AGAIN),
+        error.InputOutput => return unexpectedErrno(.IO),
+        error.IsDir => return errnoBug(.ISDIR),
+        error.NotOpenForReading => return errnoBug(.BADF),
+        error.LockViolation => return errnoBug(.AGAIN),
+        error.Unexpected => return error.Unexpected,
+    };
 }
 
-fn netWriteUnavailable(
+fn netWrite(
     userdata: ?*anyopaque,
     handle: net.Socket.Handle,
     header: []const u8,
@@ -5162,12 +5652,68 @@
     splat: usize,
 ) net.Stream.Writer.Error!usize {
     const ev: *Evented = @ptrCast(@alignCast(userdata));
-    _ = ev;
-    _ = handle;
-    _ = header;
-    _ = data;
-    _ = splat;
-    return error.NetworkDown;
+
+    var iovecs: [max_iovecs_len]iovec_const = undefined;
+    var msg: linux.msghdr_const = .{
+        .name = null,
+        .namelen = 0,
+        .iov = &iovecs,
+        .iovlen = 0,
+        .control = null,
+        .controllen = 0,
+        .flags = 0,
+    };
+    addBuf(&iovecs, &msg.iovlen, header);
+    if (data.len > 0) {
+        for (data[0 .. data.len - 1]) |bytes| addBuf(&iovecs, &msg.iovlen, bytes);
+    } else {
+        assert(splat == 0);
+    }
+
+    var splat_backup_buffer: [splat_buffer_size]u8 = undefined;
+    if (data.len > 0 and iovecs.len - msg.iovlen != 0) {
+        const pattern = data[data.len - 1];
+        switch (splat) {
+            0 => {},
+            1 => addBuf(&iovecs, &msg.iovlen, pattern),
+            else => switch (pattern.len) {
+                0 => {},
+                1 => {
+                    const splat_buffer = &splat_backup_buffer;
+                    const memset_len = @min(splat_buffer.len, splat);
+                    const buf = splat_buffer[0..memset_len];
+                    @memset(buf, pattern[0]);
+                    addBuf(&iovecs, &msg.iovlen, buf);
+                    var remaining_splat = splat - buf.len;
+                    while (remaining_splat > splat_buffer.len and iovecs.len - msg.iovlen != 0) {
+                        assert(buf.len == splat_buffer.len);
+                        addBuf(&iovecs, &msg.iovlen, splat_buffer);
+                        remaining_splat -= splat_buffer.len;
+                    }
+                    addBuf(&iovecs, &msg.iovlen, splat_buffer[0..@min(remaining_splat, splat_buffer.len)]);
+                },
+                else => for (0..@min(splat, iovecs.len - msg.iovlen)) |_| {
+                    addBuf(&iovecs, &msg.iovlen, pattern);
+                },
+            },
+        }
+    }
+
+    var cancel_region: CancelRegion = .init();
+    defer cancel_region.deinit();
+    return ev.netWriteMsg(&cancel_region, handle, &msg, linux.MSG.NOSIGNAL) catch |err| switch (err) {
+        error.Canceled => return error.Canceled,
+        error.FastOpenAlreadyInProgress => return error.FastOpenAlreadyInProgress,
+        error.ConnectionResetByPeer => return error.ConnectionResetByPeer,
+        error.SystemResources => return error.SystemResources,
+        error.NetworkUnreachable => return error.NetworkUnreachable,
+        error.HostUnreachable => return error.HostUnreachable,
+        error.NetworkDown => return error.NetworkDown,
+        error.ConnectionRefused => return error.ConnectionRefused,
+        error.AddressFamilyUnsupported => return error.AddressFamilyUnsupported,
+        error.SocketUnconnected => return error.SocketUnconnected,
+        else => |e| return e,
+    };
 }
 
 fn netWriteFileUnavailable(
@@ -5308,6 +5854,59 @@
     }
 }
 
+fn connect(
+    ev: *Evented,
+    cancel_region: *CancelRegion,
+    socket_fd: fd_t,
+    addr: *const linux.sockaddr,
+    addr_len: linux.socklen_t,
+) net.IpAddress.ConnectError!void {
+    while (true) {
+        const thread = try cancel_region.awaitIoUring();
+        thread.enqueue().* = .{
+            .opcode = .CONNECT,
+            .flags = 0,
+            .ioprio = 0,
+            .fd = socket_fd,
+            .off = addr_len,
+            .addr = @intFromPtr(addr),
+            .len = 0,
+            .rw_flags = 0,
+            .user_data = @intFromPtr(cancel_region.fiber),
+            .buf_index = 0,
+            .personality = 0,
+            .splice_fd_in = 0,
+            .addr3 = 0,
+            .resv = 0,
+        };
+        ev.yield(null, .nothing);
+        switch (cancel_region.errno()) {
+            .SUCCESS => return,
+            .INTR, .CANCELED => {},
+            .ADDRNOTAVAIL => return error.AddressUnavailable,
+            .AFNOSUPPORT => return error.AddressFamilyUnsupported,
+            .AGAIN, .INPROGRESS => |err| return errnoBug(err), // io_uring handles EINPROGRESS via FAST_POLL.
+            .ALREADY => |err| return errnoBug(err), // Previous connect should have completed via CQE.
+            .CONNREFUSED => return error.ConnectionRefused,
+            .CONNRESET => return error.ConnectionResetByPeer,
+            .HOSTUNREACH => return error.HostUnreachable,
+            .NETUNREACH => return error.NetworkUnreachable,
+            .TIMEDOUT => return error.Timeout,
+            .ACCES => return error.AccessDenied,
+            .NETDOWN => return error.NetworkDown,
+            .BADF => |err| return errnoBug(err),
+            .CONNABORTED => |err| return errnoBug(err),
+            .FAULT => |err| return errnoBug(err),
+            .ISCONN => |err| return errnoBug(err),
+            .NOENT => |err| return errnoBug(err),
+            .NOTSOCK => |err| return errnoBug(err),
+            .PERM => |err| return errnoBug(err),
+            .PROTOTYPE => |err| return errnoBug(err),
+            else => |err| return unexpectedErrno(err),
+        }
+    }
+}
+
 fn chdir(sync: *CancelRegion.Sync, path: [*:0]const u8) ChdirError!void {
     while (true) {
         try sync.cancel_region.await(.nothing);
@@ -5533,6 +6132,43 @@
     }
 }
 
+fn listen(
+    ev: *Evented,
+    cancel_region: *CancelRegion,
+    socket_fd: fd_t,
+    backlog: u31,
+) net.IpAddress.ListenError!void {
+    while (true) {
+        const thread = try cancel_region.awaitIoUring();
+        thread.enqueue().* = .{
+            .opcode = .LISTEN,
+            .flags = 0,
+            .ioprio = 0,
+            .fd = socket_fd,
+            .off = 0,
+            .addr = 0,
+            .len = backlog,
+            .rw_flags = 0,
+            .user_data = @intFromPtr(cancel_region.fiber),
+            .buf_index = 0,
+            .personality = 0,
+            .splice_fd_in = 0,
+            .addr3 = 0,
+            .resv = 0,
+        };
+        ev.yield(null, .nothing);
+        switch (cancel_region.errno()) {
+            .SUCCESS => return,
+            .INTR, .CANCELED => {},
+            .ADDRINUSE => return error.AddressInUse,
+            .BADF => |err| return errnoBug(err),
+            .NOTSOCK => |err| return errnoBug(err),
+            .OPNOTSUPP => |err| return errnoBug(err),
+            else => |err| return unexpectedErrno(err),
+        }
+    }
+}
+
 fn linkat(
     ev: *Evented,
     cancel_region: *CancelRegion,
@@ -5688,6 +6324,8 @@
 ) File.Reader.Error!usize {
     if (iov.len == 0) return 0;
     const gather = iov.len > 1 or iov[0].len > 0xfffff000;
+    var requested_bytes: usize = 0;
+    for (iov) |v| requested_bytes += v.len;
     while (true) {
         const thread = try cancel_region.awaitIoUring();
         thread.enqueue().* = .{
@@ -5709,12 +6347,15 @@
         ev.yield(null, .nothing);
         const completion = cancel_region.completion();
         switch (completion.errno()) {
-            .SUCCESS => return @as(u32, @bitCast(completion.result)),
+            .SUCCESS => {
+                const result: u32 = @bitCast(completion.result);
+                return result;
+            },
             .INTR, .CANCELED => {},
             .INVAL => |err| return errnoBug(err),
             .FAULT => |err| return errnoBug(err),
             .AGAIN => return error.WouldBlock,
-            .BADF => |err| return errnoBug(err), // File descriptor used after closed
+            .BADF => |err| return errnoBug(err),
             .IO => return error.InputOutput,
             .ISDIR => return error.IsDir,
             .NOBUFS => return error.SystemResources,
